### PR TITLE
File Search Bar

### LINKED ISSUE #3503: Searching for files
### Is there an existing feature request for this?

- [X] I have searched the existing issues before opening this feature request.

### Describe the feature you would like to see.

Searching through files of a server, just like you can with servers.

### Describe the solution you'd like.

A button in the files area that you can click to bring a pop up menu where you can enter a file name.

### Additional context to this request.

_No response_

### PR DESCRIPTION
**Pull request by mrxbox98**
This pr adds a search bar to the file manager.

closes https://github.com/pterodactyl/panel/issues/3503

### CODE REVIEW

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 1]
```diff
@@ -41,22 +50,30 @@ export default () => {
         clearFlashes('files');
         setSelectedFiles([]);
         setDirectory(hashToPath(hash));
-    }, [ hash ]);
+    }, [hash]);
```
Comment (Jelcoo) [inline resources/scripts/components/server/files/FileManagerContainer.tsx:50]: Add back the spaces

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 2]
```diff
@@ -41,22 +50,30 @@ export default () => {
...
 
     useEffect(() => {
         mutate();
-    }, [ directory ]);
+    }, [directory]);
```
Comment (Jelcoo) [inline resources/scripts/components/server/files/FileManagerContainer.tsx:54]: Add back the spaces

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 3]
```diff
@@ -41,22 +50,30 @@ export default () => {
...
 
     const onSelectAllClick = (e: React.ChangeEvent<HTMLInputElement>) => {
         setSelectedFiles(e.currentTarget.checked ? (files?.map(file => file.name) || []) : []);
     };
 
     if (error) {
         return (
-            <ServerError message={httpErrorToHuman(error)} onRetry={() => mutate()}/>
+            <ServerError message={httpErrorToHuman(error)} onRetry={() => mutate()} />
```
Comment (Jelcoo) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: Remove the space

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 4]
```diff
@@ -72,11 +89,14 @@ export default () => {
                         }
                     />
                 </ErrorBoundary>
+
+                <input onChange={searchFiles} css={tw`w-full rounded-lg bg-neutral-700 border-2 border-cyan-700 mx-10 px-4`} placeholder='search' style={{borderColor: 'rgb(59 130 246);'}}></input>
```
Comment (Jelcoo) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: Use the Field component

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 5]
```diff
@@ -72,11 +89,14 @@ export default () => {
...
+
                 <Can action={'file.create'}>
                     <ErrorBoundary>
                         <div css={tw`flex flex-shrink-0 flex-wrap-reverse md:flex-nowrap justify-end mb-4 md:mb-0 ml-0 md:ml-auto`}>
-                            <NewDirectoryButton css={tw`w-full flex-none mt-4 sm:mt-0 sm:w-auto sm:mr-4`}/>
-                            <UploadButton css={tw`flex-1 mr-4 sm:flex-none sm:mt-0`}/>
+                            <NewDirectoryButton css={tw`w-full flex-none mt-4 sm:mt-0 sm:w-auto sm:mr-4`} />
```
Comment (Jelcoo) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: Remove space

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 6]
```diff
@@ -72,11 +89,14 @@ export default () => {
...
+                            <UploadButton css={tw`flex-1 mr-4 sm:flex-none sm:mt-0`} />
```
Comment (Jelcoo) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: Remove space

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 7]
```diff
@@ -91,7 +111,7 @@ export default () => {
             </div>
             {
                 !files ?
-                    <Spinner size={'large'} centered/>
+                    <Spinner size={'large'} centered />
```
Comment (Jelcoo) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: Remove space

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 8]
```diff
@@ -102,19 +122,19 @@ export default () => {
...
+                                            <p css={tw`text-yellow-900 text-sm text-center`}>
+                                                This directory is too large to display in the browser,
+                                                limiting the output to the first 250 files.
+                                            </p>
+                                        </div>
                                     }
                                     {
                                         sortFiles(files.slice(0, 250)).map(file => (
-                                            <FileObjectRow key={file.key} file={file}/>
+                                            <FileObjectRow key={file.key} file={file} />
```
Comment (Jelcoo) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: Remove space

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 9]
```diff
@@ -102,19 +122,19 @@ export default () => {
...
                                         ))
                                     }
-                                    <MassActionsBar/>
+                                    <MassActionsBar />
```
Comment (Jelcoo) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: Remove space

**Review by Jelcoo — COMMENTED** [review]
Please fix some things

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 4 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx]: I tried using the field component in an earlier commit and the CSS spacing got all weird. When I tried setting any CSS elements on the field to fix it the input stopped working.

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 2 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx:54]: Fixed in https://github.com/pterodactyl/panel/pull/3926/commits/d09749a8dc1812985035c2e249de4a6ff6b1effd

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 1 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx:50]: Fixed in https://github.com/pterodactyl/panel/pull/3926/commits/d09749a8dc1812985035c2e249de4a6ff6b1effd

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 3 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx]: Fixed in https://github.com/pterodactyl/panel/pull/3926/commits/9862de5fd97dade3f9783d4d07a451504e830143

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 5 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx]: Fixed in https://github.com/pterodactyl/panel/pull/3926/commits/9862de5fd97dade3f9783d4d07a451504e830143

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 6 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx]: Fixed in https://github.com/pterodactyl/panel/pull/3926/commits/9862de5fd97dade3f9783d4d07a451504e830143

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 7 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx]: Fixed in https://github.com/pterodactyl/panel/pull/3926/commits/9862de5fd97dade3f9783d4d07a451504e830143

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 8 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx]: Fixed in https://github.com/pterodactyl/panel/pull/3926/commits/9862de5fd97dade3f9783d4d07a451504e830143

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 9 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx]: Fixed in https://github.com/pterodactyl/panel/pull/3926/commits/9862de5fd97dade3f9783d4d07a451504e830143

**Comment by Jelcoo** [comment]
Implementation is not very good. It covers directory names, and it also can't search in directories
firefox_sZU6YRgz0S

**Comment by mrxbox98** [comment]
> Implementation is not very good. It covers directory names, and it also can't search in directories firefox_sZU6YRgz0S

Directory name issue fixed in https://github.com/pterodactyl/panel/pull/3926/commits/43d44da1234f38eaaa77a1853ce3d757dc8cde4d. Also I am able to search in directories.

File: resources/scripts/components/server/files/FileManagerContainer.tsx (hunk 9 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerContainer.tsx]: Fixed in https://github.com/pterodactyl/panel/commit/9862de5fd97dade3f9783d4d07a451504e830143

**Comment by Jelcoo** [comment]
You can search when you're inside a folder, but not when you're in the main directory and want to search in other directories. I was the one that made the issue about this feature request, and that is the way I had in mind.

**Comment by mrxbox98** [comment]
> You can search when you're inside a folder, but not when you're in the main directory and want to search in other directories. I was the one that made the issue about this feature request, and that is the way I had in mind.

From what I can tell this is only possible with modifications to wings as only the files/folders the user is viewing are sent by wings.

**Comment by yesBad** [comment]
How does it look on mobile?

**Comment by mrxbox98** [comment]
> How does it look on mobile?

Like this
Screenshot from 2022-02-04 07-14-46

**Comment by FoksVHox** [comment]
Does the searchbar still overlap the filestructurr breadcrumbs, on desktop?

**Comment by mrxbox98** [comment]
> Does the searchbar still overlap the filestructurr breadcrumbs, on desktop?

Nope

**Comment by FoksVHox** [comment]
> > Does the searchbar still overlap the filestructurr breadcrumbs, on desktop?
> 
> 
> 
> Nope

Glad to hear.

I however feel that the mobile version should have a little more work done, so that it matches with the quality of the rest of the ui

File: resources/scripts/components/server/files/FileManagerBreadcrumbs.tsx [hunk 10]
```diff
@@ -36,7 +36,7 @@ export default ({ renderLeft, withinFileEditor, isNewFile }: Props) => {
         });
 
     return (
-        <div css={tw`flex flex-grow-0 items-center text-sm text-neutral-500 overflow-x-hidden`}>
+        <div css={tw`flex flex-grow-0 items-center text-sm text-neutral-500`}>
```
Comment (DaneEveritt) [inline resources/scripts/components/server/files/FileManagerBreadcrumbs.tsx]: Why was this change made?

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 11]
```diff
@@ -21,9 +21,11 @@ import ErrorBoundary from '@/components/elements/ErrorBoundary';
 import { FileActionCheckbox } from '@/components/server/files/SelectFileCheckbox';
 import { hashToPath } from '@/helpers';
 
+let searchString = '';
```
Comment (DaneEveritt) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: This is not a correct way of handling component states.

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 12]
```diff
@@ -21,9 +21,11 @@ import ErrorBoundary from '@/components/elements/ErrorBoundary';
...
+
 const sortFiles = (files: FileObject[]): FileObject[] => {
     const sortedFiles: FileObject[] = files.sort((a, b) => a.name.localeCompare(b.name)).sort((a, b) => a.isFile === b.isFile ? 0 : (a.isFile ? 1 : -1));
-    return sortedFiles.filter((file, index) => index === 0 || file.name !== sortedFiles[index - 1].name);
+    return sortedFiles.filter((file, index) => index === 0 || file.name !== sortedFiles[index - 1].name).filter((file) => file.name.includes(searchString));
```
Comment (DaneEveritt) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: This is a case sensitive search when done this way.

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 13]
```diff
@@ -71,12 +82,20 @@ export default () => {
                             />
                         }
                     />
+                    <input
+                        onChange={searchFiles}
+                        css={tw`rounded-lg bg-neutral-700 border-2 border-cyan-700 md:mx-6 w-full px-4 mb-4 md:mb-0`}
+                        placeholder='search'
```
Comment (DaneEveritt) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: This should be capitalized.

File: resources/scripts/components/server/files/FileManagerContainer.tsx [hunk 14]
```diff
@@ -71,12 +82,20 @@ export default () => {
...
+                        style={{ borderColor: 'rgb(59 130 246);' }}
```
Comment (DaneEveritt) [inline resources/scripts/components/server/files/FileManagerContainer.tsx]: This shouldn't be inline styled, in addition to this input bar not matching the rest of the application's inputs.

**Review by DaneEveritt — CHANGES_REQUESTED** [review]

**Comment by mrxbox98** [comment]
Since v2 is coming next I will change the target of this PR to v2 soon

File: resources/scripts/components/server/files/FileManagerBreadcrumbs.tsx (hunk 10 above)
Comment (mrxbox98) [reply resources/scripts/components/server/files/FileManagerBreadcrumbs.tsx]: The directory path would be masked by the input field if ``overflow-x-hidden`` was present.

**Comment by mrxbox98** [comment]
Updated UI:
Regular Desktop

Mobile:
iPhone XR

**Comment by tmunsch** [comment]
Nice work!

**Comment by mrxbox98** [comment]
@matthewpi Could you review this when you have time

**Pull request closed without merging**